ledger/src/blockstore.rs

```diff
@@ -1921,6 +1922,127 @@ impl Blockstore {
+                            EncodedTransaction::Json(ui_transaction) => {
+                                match ui_transaction.message {
+                                    UiMessage::Raw(message) => {
+                                        let address = address.to_string();
+                                        if !message
+                                            .account_keys
+                                            .iter()
+                                            .any(|account_address| *account_address == address)
+                                        {
+                                            return Err(BlockstoreError::IO(IOError::new(
+                                                ErrorKind::Other,
+                                                format!(
+                                                    "Invalid start_after signature: {}",
+                                                    start_after
+                                                ),
+                                            )));
+                                        }
+                                    }
+                                    _ => {
+                                        // Should never happen...
+                                        return Err(BlockstoreError::IO(IOError::new(
+                                            ErrorKind::Other,
+                                            "Unexpected transaction message encoding".to_string(),
+                                        )));
+                                    }
+                                }
+                            }
+                            _ => {
+                                // Should never happen...
+                                return Err(BlockstoreError::IO(IOError::new(
+                                    ErrorKind::Other,
+                                    "Unexpected transaction encoding".to_string(),
+                                )));
+                            }
+                        }
+                        (slot, Some(start_after))
+                    }
+                }
+            }
+        };
+
+        // Fetch the list of signatures that affect the given address
+        let first_available_block = self.get_first_available_block()?;
+        let mut address_signatures = vec![];
+        loop {
+            if address_signatures.len() >= limit {
+                address_signatures.truncate(limit);
+                break;
+            }
+
+            let mut signatures = self.find_address_signatures(address, slot, slot)?;
+            if let Some(start_after) = start_after {
+                if let Some(index) = signatures
+                    .iter()
+                    .position(|(_slot, signature)| *signature == start_after)
+                {
+                    address_signatures.append(&mut signatures.split_off(index + 1));
+                    signatures.clear();
+                } else {
```

Comment:
I don't think this part does quite what you hope. `AddressSignatures` are indexed by primary_index-address-slot-signature. This means that within a slot, signatures are stored in order of signature bytes, not in the order they were processed. So here you're splitting off the signatures that happen to have a larger first byte than `start_after`.

The best we could do here is grab all the signatures in the same block as the `start_after` sig (in which case, we only need the slot, and don't need the requirement that `start_after` is from a transaction that contains `address`). Then we'd need to deserialize that block (`Blockstore::get_entries_in_data_block()`) to determine which AddressSignatures are after the `start_after` sig. Does that make sense? I could mock up a commit that does this.

Reply:
@CriesofCarrots - thanks!  I just pushed a new commit that reworks the function.  Can you please take a look now?